awx/main/scheduler/task_manager.py

```diff
@@ -720,6 +726,28 @@ def record_aggregate_metrics_and_exit(self, *args):
         self.record_aggregate_metrics()
         sys.exit(1)
 
+    def set_start_task_limit(self):
+        # If we are not making any progress AND we are hitting the start_task_limit,
+        # we need to increase the number of jobs we are considering
+        unstarted_ct = self.pending_tasks_considered_ct - self.tasks_started_ct
+        new_start_task_limit = settings.START_TASK_LIMIT + unstarted_ct
+        if new_start_task_limit <= settings.START_TASK_LIMIT:
+            logger.info(f'Task manager started {self.tasks_started_ct} of {self.pending_tasks_considered_ct} pending jobs, ' 'removing custom start_task_limit')
+            self.redis.delete('start_task_limit')
+        elif new_start_task_limit != self.start_task_limit:
+            msg = (
+                f'Task manager started {self.tasks_started_ct} of {self.pending_tasks_considered_ct} pending jobs '
+                f'with limit of {self.start_task_limit}, next limit will be {new_start_task_limit}'
+            )
+            if self.tasks_started_ct == 0:
+                logger.warning(msg)
+            else:
+                logger.info(msg)
+            self.redis.set('start_task_limit', new_start_task_limit)
+
+        if new_start_task_limit > self.start_task_limit:
```

Comment:
This criteria is a fancy way of saying "if we _actually processed_ fewer jobs than the task_start_limit, then reschedule". In that case, we would also set a higher value for the task_start_limit.